rfcs/002 - Custom Module Functions RFC.md

```diff
@@ -0,0 +1,80 @@
+
+This has always been a sore point architecturally as encapsulation of information is broken, and it is not truly possible to adjust the behavior of the functions in question without affecting prior module versions at the same time. It is also very difficult (though not impossible) to introduce new functions because 4 different files across the repository must be updated in sync to do so correctly, and can only be done in a way that would be considered a SemVer minor update, major is completely not allowed.
+
+It is proposed to deprecate/drop the `iasql_operation` entity and associated enum, and create a new graphile worker job type, called `rpc` and a new entity named `iasql_rpc` with no enum. Instead two fields would be added as an "addressing system" of module name and method name, which the module would implement within an `rpc` sub-object where all properties are the method names and the values are async functions that take a `Context` object as the first argument and then any arguments from the SQL side are added as follow-up parameters, all as strings similar to the current `params` array.
+
+The `rpc` job in the `scheduler.ts` file would attempt to find and execute this function within the `rpc` object of the named module, and provide relevant error responses if it could not find the function or the function itself failed in some way or the other. Because this is now generalized, it is also possible to have the Module logic in `interfaces.ts` automatically generate "bindings" in Postgres for these functions without requiring hand-written SQL at all.
+
+This would also allow us to move the majority of the `src/services/iasql.ts` functions into the `iasql_functions` module, allowing behavior changes to happen between version bumps of IaSQL and properly segmenting scope again. We would also [tackle this task](https://github.com/iasql/iasql-engine/issues/1016) on the new `iasql_rpc` entity instead of the `iasql_operation` entity.
+
+On a similar note, the current hand-editing of the migration files to insert custom pure-SQL functions is error prone when the migration is regenerated on entity changes. I propose that we have a similar `sql` sub-type on the modules similar to the old `migration` type that automatically runs whatever custom SQL was given to it after the migration up and before the migration down, to prevent this, treating it sort of like a second migration step, but more ergonomic than having to hand write a TypeORM migration file.
```

Comment:
Yes, this is necessary. It would be something like
```ts
sql: {
  afterMigrationUp,
  beforeMigrationDown
}
```

Reply:
Yep :) Though I would probably use terminology like `afterInstall` and `beforeUninstall`